pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -84,3 +85,59 @@ func ValidateObjectParamRequiredKeys(pipelineParameters []v1beta1.ParamSpec, pip
 
 	return nil
 }
+
+// ValidateParameterTypesInMatrix validates the type of Parameter for Matrix.Params
+// and Matrix.Include.Params after any replacements are made from Task parameters or results
+// Matrix.Params must be of type array. Matrix.Include.Params must be of type string.
+func ValidateParameterTypesInMatrix(state PipelineRunState) error {
+	for _, rpt := range state {
+		m := rpt.PipelineTask.Matrix
+		if m.HasInclude() {
+			for _, include := range m.Include {
+				for _, param := range include.Params {
+					if param.Value.Type != v1beta1.ParamTypeString {
+						return fmt.Errorf("parameters of type string only are allowed, but got param type %s", string(param.Value.Type))
+					}
+				}
+			}
+		}
+		if m.HasParams() {
+			for _, param := range m.Params {
+				if param.Value.Type != v1beta1.ParamTypeArray {
+					return fmt.Errorf("parameters of type array only are allowed, but got param type %s", string(param.Value.Type))
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// ValidateUniqueParamsInMatrix validates there are no duplicate parameters
+// within Matrix.Params or Matrix.Include Params
+func ValidateUniqueParamsInMatrix(state PipelineRunState) error {
+	var errs error
+	for _, rpt := range state {
+		m := rpt.PipelineTask.Matrix
+		if m.HasInclude() {
+			for _, include := range m.Include {
+				errs = validateDuplicateParameters(include.Params, "matrix.include.params")
+			}
+		}
+		if m.HasParams() {
+			errs = validateDuplicateParameters(m.Params, "matrix.params")
+		}
+	}
+	return errs
+}
+
+// validateDuplicateParameters checks if a parameter with the same name is defined more than once
+func validateDuplicateParameters(ps v1beta1.Params, typeParam string) (errs error) {
+	taskParamNames := sets.NewString()
+	for i, param := range ps {
+		if taskParamNames.Has(param.Name) {
+			errs = fmt.Errorf(fmt.Sprintf("parameter names must be unique, the parameter %s is also defined at %s[%d].name", param.Name, typeParam, i))
```

Comment:
This looks like it only returns the last error. Is the intention to return a multierror or just the last one?